### PR TITLE
Fixes part of #59: Testing module builds with both Bazel + Gradle [BLOCKED: #1482]

### PR DESCRIPTION
**Pull request by miaboloix**
## Explanation
Fixes part of #59: Testing module builds with both Bazel + Gradle
I added the `/testing` module BUILD file and wrote a `kt_android_library()` rule that builds the source files in `/testing`. I also configured Robolectric testing for this module using custom macros `oppia_android_testing()` and `testing_tests()`.

To build the `/testing` app module in Bazel: `bazel build //testing`
To run all tests in the module: `bazel test //testing/...`

Notes: 
1. Gradle seems to fail when Bazel build artifacts are present. Be sure to remove them before building Gradle by running `bazel clean`
2. This PR depends on  #1482

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [ ] The PR's branch is based on "develop" and not on any other branch. 
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @miaboloix! Does this build without issues? I realized that this module doesn't actually have any UI resources unfortunately, so the potential Android resource building issues will need to wait until we start building the app module.

**Comment by miaboloix** [comment]
> Thanks @miaboloix! Does this build without issues? I realized that this module doesn't actually have any UI resources unfortunately, so the potential Android resource building issues will need to wait until we start building the app module.

Yes, this builds!

**Comment by BenHenning** [comment]
@anandwana001 can you take a first pass on this? I'll take a closer look on Monday.

**Comment by anandwana001** [comment]
> @anandwana001 can you take a first pass on this? I'll take a closer look on Monday.

Yes, Sure

File: testing/BUILD.bazel [hunk 1]
```diff
@@ -7,16 +7,65 @@ package(default_visibility = ["//visibility:public"])
...
+         artifact("com.crashlytics.sdk.android:crashlytics:2.9.8"),
+         artifact("io.fabric.sdk.android:fabric:1.4.7"),
+         artifact("com.google.gms:google-services:4.3.3"),
+         artifact("com.android.support:support-annotations"),
+         artifact("androidx.test.ext:junit"),
+         artifact("org.mockito:mockito-core:2.19.0"),
      ],
  )
 
+
```
Comment (anandwana001) [inline testing/BUILD.bazel:35]: We can remove 1 empty line.

**Review by anandwana001 — CHANGES_REQUESTED** [review]
I failed to build this. I will try again as it was time out issue. Let me go through other or related to Bazel and will get back to this.

**Comment by anandwana001** [comment]
@miaboloix I tried again and getting a bit difficulty in a successful build with the below error

```
 @robolectric//bazel:android-all-jars depends on @org_robolectric_android_all_6_0_1_r3_robolectric_r1//jar:jar in repository @org_robolectric_android_all_6_0_1_r3_robolectric_r1 which failed to fetch. no such package '@org_robolectric_android_all_6_0_1_r3_robolectric_r1//jar': java.io.IOException: Error downloading [https://repo1.maven.org/maven2/org/robolectric/android-all/6.0.1_r3-robolectric-r1/android-all-6.0.1_r3-robolectric-r1.jar, https://jcenter.bintray.com/org/robolectric/android-all/6.0.1_r3-robolectric-r1/android-all-6.0.1_r3-robolectric-r1.jar] to /private/var/tmp/_bazel_akshaynandwana/86e7652338e118cae6d5283ced282727/external/org_robolectric_android_all_6_0_1_r3_robolectric_r1/android-all-6.0.1_r3-robolectric-r1.jar: Unknown host: jcenter.bintray.com
```

**Comment by miaboloix** [comment]
> @miaboloix I tried again and getting a bit difficulty in a successful build with the below error
> 
> ```
>  @robolectric//bazel:android-all-jars depends on @org_robolectric_android_all_6_0_1_r3_robolectric_r1//jar:jar in repository @org_robolectric_android_all_6_0_1_r3_robolectric_r1 which failed to fetch. no such package '@org_robolectric_android_all_6_0_1_r3_robolectric_r1//jar': java.io.IOException: Error downloading [https://repo1.maven.org/maven2/org/robolectric/android-all/6.0.1_r3-robolectric-r1/android-all-6.0.1_r3-robolectric-r1.jar, https://jcenter.bintray.com/org/robolectric/android-all/6.0.1_r3-robolectric-r1/android-all-6.0.1_r3-robolectric-r1.jar] to /private/var/tmp/_bazel_akshaynandwana/86e7652338e118cae6d5283ced282727/external/org_robolectric_android_all_6_0_1_r3_robolectric_r1/android-all-6.0.1_r3-robolectric-r1.jar: Unknown host: jcenter.bintray.com
> ```

What exactly were you building? `testing_main_lib` or `testing_test_lib`? I haven't been able to reproduce this error. 

I just pushed a change that removed one of the unused robolectric dependencies from `testing_main_lib` -- can you see if that did anything? Also, if it is still failing, you may be able to get a more in-depth error message by building `testing_main_lib_kt` or `testing_test_lib_kt`.

File: testing/BUILD.bazel (hunk 1 above)
Comment (miaboloix) [reply testing/BUILD.bazel:35]: Done!

File: WORKSPACE [hunk 2]
```diff
@@ -104,7 +102,7 @@ load("@tools_android//tools/googleservices:defs.bzl", "google_services_workspace
 google_services_workspace_dependencies()
 
 load("@rules_jvm_external//:defs.bzl", "maven_install")
-# TODO: Remove unused dependencies once android_local_test errors are resolved
+#TODO: Remove unused dependencies once android_local_test errors are resolved
```
Comment (BenHenning) [inline WORKSPACE:105]: Please resolve.

Comment (BenHenning) [inline WORKSPACE]: Please resolve this TODO.

File: WORKSPACE [hunk 3]
```diff
@@ -22,8 +22,6 @@ http_archive(
     url = "https://github.com/bazelbuild/rules_jvm_external/archive/%s.zip" % RULES_JVM_EXTERNAL_TAG,
 )
 
-load("@rules_jvm_external//:defs.bzl", "maven_install")
```
Comment (BenHenning) [inline WORKSPACE]: Should this be removed? It seems like it’s still being used.

File: testing/BUILD.bazel [hunk 4]
```diff
@@ -7,16 +8,60 @@ package(default_visibility = ["//visibility:public"])
...
+         artifact("io.fabric.sdk.android:fabric:1.4.7"),
+         artifact("com.google.gms:google-services:4.3.3"),
+         artifact("com.android.support:support-annotations"),
+         artifact("androidx.test.ext:junit"),
+         artifact("org.mockito:mockito-core:2.19.0"),
      ],
- )
+)
+
+# TODO: Add android_local_test rule
```
Comment (BenHenning) [inline testing/BUILD.bazel]: Please resolve.

File: testing/BUILD.bazel [hunk 5]
```diff
@@ -7,16 +8,60 @@ package(default_visibility = ["//visibility:public"])
 kt_android_library(
      name = "testing_main_lib",
```
Comment (BenHenning) [inline testing/BUILD.bazel]: Prefer not setting the default visibility for a package. We should lock down access so that only the things that need to be public are made public.

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @miaboloix! Overall this LGTM but it partly depends on upstream changes being done, and does depend on the test TODO being resolved.

File: testing/BUILD.bazel (hunk 5 above)
Comment (miaboloix) [reply testing/BUILD.bazel]: Done!

File: testing/BUILD.bazel (hunk 4 above)
Comment (miaboloix) [reply testing/BUILD.bazel]: okay!

File: WORKSPACE (hunk 2 above)
Comment (miaboloix) [reply WORKSPACE]: Done!

**Comment by miaboloix** [comment]
Note: I realized that Robolectric testing for this module was not dependent on anything not already configured in utility - so I moved the changes from #1495 that applied to the testing module here.

File: oppia_android_test.bzl [hunk 6]
```diff
@@ -0,0 +1,45 @@
+load("@rules_jvm_external//:defs.bzl", "artifact")
+load("@io_bazel_rules_kotlin//kotlin:kotlin.bzl", "kt_android_library")
+
+def oppia_android_test(name, srcs, test_manifest, custom_package, resource_files,
+                       test_class, src_library_name):
+  '''
+  This macro exists as a way to set up a test in Oppia Android to be run with Bazel.
+  This macro creates a library for an individual test that is fed as a dependency into an
+  android_local_test() rule which configures the test to be run with Roboletric and Bazel.
```
Comment (BenHenning) [inline oppia_android_test.bzl]:
```suggestion
  Creates an Oppia test target for running the specified test as an Android local test with Kotlin support. Note that this creates an additional, internal library.
```

My suggestion needs wrapping, but suggestion to simplify the wording a tad. (Note that specifying Bazel in documentation is unnecessary in most cases--Bazel is implied since this code lives in Bazel files).

File: oppia_android_test.bzl [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+  Args:
+      name: str. The name of the Kotlin test file without the '.kt' suffix.
+      srcs: str. The name of the Kotlin test file to be run.
```
Comment (BenHenning) [inline oppia_android_test.bzl]: Is this correct? Isn't this supposed to be a list of Kotlin source files that include the test file being run?

File: oppia_android_test.bzl [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+      test_manifest: str. The path to the test manifest file.
+      custom_package: str. The module's package. Example: 'org.oppia.utility'.
+      resource_files: str. The path to the resource files. This is typically a glob([]).
+      test_class: The package of the src file. For example, if the src is 'FakeEventLoggerTest.kt',
+          then the test_class would be "org.oppia.testing.FakeEventLoggerTest".
+      src_library_name: str. The name of the library that builds the module's test files.
```
Comment (BenHenning) [inline oppia_android_test.bzl]: Why do we need this? Can't we just pass in the deps that the test needs & build each test by itself via its test library below? The additional combined tests libraries seem unnecessary.

I was assuming that this macro would essentially be a final test macro we would use into & post phase 2, whereas the separate ``domain_test``-esque macros are temporary in that they simplify the test definitions (maybe even sharing a deps list).

File: oppia_android_test.bzl [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+  native.android_local_test(
+    name = name,
+    custom_package = custom_package,
+    test_class = test_class,
+    manifest = test_manifest,
+    deps = [
+       ":" + name + "_lib",
+       ":dagger",
+       "@robolectric//bazel:android-all",
+       artifact("org.jetbrains.kotlin:kotlin-reflect"),
```
Comment (BenHenning) [inline oppia_android_test.bzl]: This isn't specifically required for all tests--prefer to define it as a dep per test target, instead.

File: oppia_android_test.bzl [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+  )
+
+  native.android_local_test(
+    name = name,
+    custom_package = custom_package,
+    test_class = test_class,
+    manifest = test_manifest,
+    deps = [
+       ":" + name + "_lib",
+       ":dagger",
```
Comment (BenHenning) [inline oppia_android_test.bzl]: Should be defined as an additional dep passed into this macro, not assumed present for all tests.

File: testing/src/test/AndroidManifest.xml [hunk 11]
```diff
@@ -0,0 +1,5 @@
+<manifest xmlns:android="http://schemas.android.com/apk/res/android"
+  package="org.oppia.testing">
+  <uses-sdk android:minSdkVersion="19"
+    android:maxSdkVersion="28" />
```
Comment (BenHenning) [inline testing/src/test/AndroidManifest.xml]: Prefer target, also prefer targeting 29.

**Review by BenHenning — CHANGES_REQUESTED** [review]
Primarily focused on ``oppia_android_test``. I think my comments here will affect most of your other PRs.

**Comment by anandwana001** [comment]
> > @miaboloix I tried again and getting a bit difficulty in a successful build with the below error
> > ```
> >  @robolectric//bazel:android-all-jars depends on @org_robolectric_android_all_6_0_1_r3_robolectric_r1//jar:jar in repository @org_robolectric_android_all_6_0_1_r3_robolectric_r1 which failed to fetch. no such package '@org_robolectric_android_all_6_0_1_r3_robolectric_r1//jar': java.io.IOException: Error downloading [https://repo1.maven.org/maven2/org/robolectric/android-all/6.0.1_r3-robolectric-r1/android-all-6.0.1_r3-robolectric-r1.jar, https://jcenter.bintray.com/org/robolectric/android-all/6.0.1_r3-robolectric-r1/android-all-6.0.1_r3-robolectric-r1.jar] to /private/var/tmp/_bazel_akshaynandwana/86e7652338e118cae6d5283ced282727/external/org_robolectric_android_all_6_0_1_r3_robolectric_r1/android-all-6.0.1_r3-robolectric-r1.jar: Unknown host: jcenter.bintray.com
> > ```
> 
> What exactly were you building? `testing_main_lib` or `testing_test_lib`? I haven't been able to reproduce this error.
> 
> I just pushed a change that removed one of the unused robolectric dependencies from `testing_main_lib` -- can you see if that did anything? Also, if it is still failing, you may be able to get a more in-depth error message by building `testing_main_lib_kt` or `testing_test_lib_kt`.

Could you please add How to build description in PR description on the basis of `oppia_android_test.bzl`?

File: oppia_android_test.bzl (hunk 6 above)
Comment (miaboloix) [reply oppia_android_test.bzl]: Okay!

File: testing/src/test/AndroidManifest.xml (hunk 11 above)
Comment (miaboloix) [reply testing/src/test/AndroidManifest.xml]: Done!

File: oppia_android_test.bzl (hunk 8 above)
Comment (miaboloix) [reply oppia_android_test.bzl]: You are right! `src_library_name` is unnecessary! As for the dependencies, are you saying that macros like `testing_test()` and `domain_test()` should feed in module-specific dependencies into `oppia_android_test` or are you thinking that the actual calls to `testing_test()` and `domain_test()` found in the BUILD file should feed in test-specific dependencies? From what I have seen, most of the time dependencies don't vary too much from test to test within a module... but what do you think? Maybe there is value is having the list of dependencies visible from the BUILD file (in terms of readability, maybe?). For now, I have just created a general TEST_DEPS list in the BUILD file.

File: oppia_android_test.bzl (hunk 9 above)
Comment (miaboloix) [reply oppia_android_test.bzl]: Sounds good!

File: oppia_android_test.bzl (hunk 10 above)
Comment (miaboloix) [reply oppia_android_test.bzl]: Okay!

File: oppia_android_test.bzl (hunk 7 above)
Comment (miaboloix) [reply oppia_android_test.bzl]: Fixed!

**Comment by miaboloix** [comment]
> > > @miaboloix I tried again and getting a bit difficulty in a successful build with the below error
> > > ```
> > >  @robolectric//bazel:android-all-jars depends on @org_robolectric_android_all_6_0_1_r3_robolectric_r1//jar:jar in repository @org_robolectric_android_all_6_0_1_r3_robolectric_r1 which failed to fetch. no such package '@org_robolectric_android_all_6_0_1_r3_robolectric_r1//jar': java.io.IOException: Error downloading [https://repo1.maven.org/maven2/org/robolectric/android-all/6.0.1_r3-robolectric-r1/android-all-6.0.1_r3-robolectric-r1.jar, https://jcenter.bintray.com/org/robolectric/android-all/6.0.1_r3-robolectric-r1/android-all-6.0.1_r3-robolectric-r1.jar] to /private/var/tmp/_bazel_akshaynandwana/86e7652338e118cae6d5283ced282727/external/org_robolectric_android_all_6_0_1_r3_robolectric_r1/android-all-6.0.1_r3-robolectric-r1.jar: Unknown host: jcenter.bintray.com
> > > ```
> > 
> > 
> > What exactly were you building? `testing_main_lib` or `testing_test_lib`? I haven't been able to reproduce this error.
> > I just pushed a change that removed one of the unused robolectric dependencies from `testing_main_lib` -- can you see if that did anything? Also, if it is still failing, you may be able to get a more in-depth error message by building `testing_main_lib_kt` or `testing_test_lib_kt`.
> 
> Could you please add How to build description in PR description on the basis of `oppia_android_test.bzl`?

What do you mean? Whenever you run a test in this module, you are calling the custom macro `testing_test` (implemented in testing_test.bzl) which calls `oppia_android_test` (implemented in oppia_android_test.bazl). So in order to invoke `oppia_android_test.bzl` you just need to build a test :) The commands you would need are in the description of the PR. Please let me know if I misunderstood your question!

File: utility/src/main/AndroidManifest.xml [hunk 12]
```diff
@@ -1,5 +1,5 @@
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
   package="org.oppia.util">
-  <uses-sdk android:minSdkVersion="19"
+  <uses-sdk
```
Comment (miaboloix) [inline utility/src/main/AndroidManifest.xml]: Note: minSDKversion here was causing Gradle to fail.

File: WORKSPACE [hunk 13]
```diff
@@ -112,29 +112,24 @@ maven_install(
...
         "org.jetbrains.kotlinx:kotlinx-coroutines-test:1.2.2",
         "org.jetbrains.kotlin:kotlin-stdlib-jdk7:jar:1.3.72",
         "org.jetbrains.kotlin:kotlin-test-junit:1.3.72",
+        "org.jetbrains.kotlin:kotlin-reflect:1.3.41",
         "com.google.truth:truth:0.43",
         "com.google.firebase:firebase-analytics:17.4.4",
         "com.google.firebase:firebase-crashlytics:17.1.1",
-        "org.jetbrains.kotlin:kotlin-reflect:1.3.41",
+        "com.crashlytics.sdk.android:crashlytics:2.9.8",
+        "io.fabric.sdk.android:fabric:1.4.7",
```
Comment (BenHenning) [inline WORKSPACE:123]: Nit: please alphabetize these & all other deps lists when you update them.

File: utility/src/main/AndroidManifest.xml (hunk 12 above)
Comment (BenHenning) [reply utility/src/main/AndroidManifest.xml]: What was the error? Without additional context, it's hard to know what suggestions to make. We definitely need to specify a minimum otherwise Bazel will assume to use 1 which is incorrect for Oppia Android.

**Review by BenHenning — APPROVED** [review]
Thanks @miaboloix --generally LGTM, but I have some concerns about the min version issue. Let's make sure that gets resolved before merging this PR in.

File: oppia_android_test.bzl (hunk 8 above)
Comment (BenHenning) [reply oppia_android_test.bzl]: I see you chose to do the second & define the test deps at the test definitions. I think that's fine--I had the former in mind (to share the same deps list for all tests of a particular category), but both work and are reasonable. Phase 2 will have specific deps lists for each test, so both solutions will be replaced in the coming weeks.

**Review by anandwana001 — APPROVED** [review]
LGTM
Thanks @miaboloix

**Comment by anandwana001** [comment]
> > > > @miaboloix I tried again and getting a bit difficulty in a successful build with the below error
> > > > ```
> > > >  @robolectric//bazel:android-all-jars depends on @org_robolectric_android_all_6_0_1_r3_robolectric_r1//jar:jar in repository @org_robolectric_android_all_6_0_1_r3_robolectric_r1 which failed to fetch. no such package '@org_robolectric_android_all_6_0_1_r3_robolectric_r1//jar': java.io.IOException: Error downloading [https://repo1.maven.org/maven2/org/robolectric/android-all/6.0.1_r3-robolectric-r1/android-all-6.0.1_r3-robolectric-r1.jar, https://jcenter.bintray.com/org/robolectric/android-all/6.0.1_r3-robolectric-r1/android-all-6.0.1_r3-robolectric-r1.jar] to /private/var/tmp/_bazel_akshaynandwana/86e7652338e118cae6d5283ced282727/external/org_robolectric_android_all_6_0_1_r3_robolectric_r1/android-all-6.0.1_r3-robolectric-r1.jar: Unknown host: jcenter.bintray.com
> > > > ```
> > > 
> > > 
> > > What exactly were you building? `testing_main_lib` or `testing_test_lib`? I haven't been able to reproduce this error.
> > > I just pushed a change that removed one of the unused robolectric dependencies from `testing_main_lib` -- can you see if that did anything? Also, if it is still failing, you may be able to get a more in-depth error message by building `testing_main_lib_kt` or `testing_test_lib_kt`.
> > 
> > 
> > Could you please add How to build description in PR description on the basis of `oppia_android_test.bzl`?
> 
> What do you mean? Whenever you run a test in this module, you are calling the custom macro `testing_test` (implemented in testing_test.bzl) which calls `oppia_android_test` (implemented in oppia_android_test.bazl). So in order to invoke `oppia_android_test.bzl` you just need to build a test :) The commands you would need are in the description of the PR. Please let me know if I misunderstood your question!

Yes, you are correct. Earlier I think I am missing some file from my side, not fetched properly I guess. It's perfect now. Tried building and running test cases, all are working perfectly correct now. 
Thank you :)

File: utility/src/main/AndroidManifest.xml [hunk 14]
```diff
@@ -1,5 +1,5 @@
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
   package="org.oppia.util">
-  <uses-sdk
+  <uses-sdk android:minSdkVersion="19"
     android:targetSdkVersion="29" />
```
Comment (miaboloix) [inline utility/src/main/AndroidManifest.xml]: @BenHenning Strangely enough, I was able to add back the minSDK without Gradle freaking out. I think the error I was getting before may have been due to the weird Gradle/Android Studio issues I was having earlier in the week!

File: WORKSPACE (hunk 13 above)
Comment (miaboloix) [reply WORKSPACE:123]: On it!

File: utility/src/main/AndroidManifest.xml (hunk 12 above)
Comment (miaboloix) [reply utility/src/main/AndroidManifest.xml]: I was able to add this back without any issues. See my latest comment below :)

File: utility/src/main/AndroidManifest.xml (hunk 14 above)
Comment (BenHenning) [reply utility/src/main/AndroidManifest.xml]: Ah, makes sense. Thanks for re-adding ti!

**Review by BenHenning — APPROVED** [review]
Thanks!
